### DISCUSSION TITLE
how to insert RLS  to inser a group member？

### DISCUSSION THREAD

**Discussion by jonz-tech** [supabase/supabase#26646] [Questions] [answered]
I have a table name group_members like this:

![group_member_fields](https://github.com/supabase/supabase/assets/6789118/eade405b-559b-47e9-96e6-af8450ae5252)

field meaning :
gid: the group id
role: 1: admin,  2: manager, 3: member

here is my question:
I want to insert a goup member only when the authented user is group  manager or admin.

I try some policy like this:
`
create policy "CREATE POLICY "insert_based_on_user_role" ON "public"."tb_group_members"

AS PERMISSIVE FOR INSERT TO PUBLIC

WITH CHECK (

    EXISTS (
        SELECT 1
        FROM public.tb_group_members
        WHERE role in (1,2)
        AND uid = auth.uid() 
        AND gid = NEW.gid
        AND validate = 1 
    )

);
`

 but some error occur: Failed to update database policy: undefined

so, how can I write the correct RLS policy?

**Comment by GaryAustin1** [comment]
You show doing create policy  then have create policy again.  You only do that once.  Where are you entering this?

**Reply by GaryAustin1** [reply]
Also you have NEW.gid there is no NEW object in RLS. You just use the column names for the table.

Here is an example of dealing with same column names...
```
EXISTS(SELECT 1 FROM user_tenant AS ut
             WHERE ut.id = "user".id
               AND ut.tenant_id = current_setting('my_user.current_tenant')::uuid
            )
```

**Reply by jonz-tech** [reply]
![image](https://github.com/supabase/supabase/assets/6789118/431cbac6-fd54-407a-8d99-493187cf9f66)
selected a table ,the entry is in the right-top , create a new policy and choose the insert Templates

**Reply by GaryAustin1** [reply]
And another example of avoiding dealing with same column names.
Just showing the relevant part where you had NEW.   
```
(SELECT 1 FROM user_website
              WHERE user_website.website_id = website.website_id
              AND user_website.user_id = auth.uid()
             )
 ```       
 
 Not sure what you are showing in your latest post except maybe that is where you enter the policy.

**Reply by jonz-tech** [reply]
in some case of triger function, it had 'new.user_id' like this: 
`DECLARE
  existing_group_count integer;
  -- insert_gid integer;
BEGIN
  SELECT COUNT(*) INTO existing_group_count FROM public.tb_group WHERE group_name = 'Personal' AND uid = new.user_id;
  IF existing_group_count = 0 THEN
    INSERT INTO public.tb_group (group_name, uid,created_uid,validate) VALUES ('Personal',new.user_id,new.user_id,1);
  END IF;
  return new;
END;` 
link with this: https://supabase.com/docs/guides/auth/managing-user-data
i want to add a new row into the table base on the insert group id, auth uid, and role in (1,2)

**Reply by GaryAustin1** [reply]
Yes trigger functions have NEW and possibly OLD.
RLS does not.
WITH CHECK looks at then data being inserted.  You don't use NEW, normally just the column name.  But if you are using inside a select on another table and you have the same column name there and want to compare them, then you have to put table.col so it knows which one you want.
USING looks at the existing row (like OLD would in a delete).

**Reply by jonz-tech** [reply]
still confuse about it. 

if a auth user has 3 group,  he is 2 group's admin, an 1 group's member. Only admin role can add group member. 
some rows like this in group_member_table:
gid    role                  uid
21     admin              a_id
22    admin              a_id
23    member          a_id
...

so, what is ther correct RLS policy to check Using  when insert group's member  in the right role and gid ?

**Reply by GaryAustin1** [reply]
You would do something like you were showing, just don't use NEW.
```
EXISTS (
    SELECT 1
    FROM public.tb_group_members
    WHERE role in (1,2)
    AND uid = auth.uid() 
    AND gid = tb_group_members.gid
    AND validate = 1 
)
```

**Reply by jonz-tech** [reply]
sorry, i had try this. but  a recursion Error response: 

ERROR: 42P17: infinite recursion detected in policy for relation "tb_group_members" CONTEXT: SQL statement "INSERT INTO tb_group_members (gid, uid,role,created_uid, validate) VALUES (gid_input,uid_input,role_input,auth_uid,1)" PL/pgSQL function fn_member_add(integer,uuid,integer) line 38 at SQL statement

**Reply by GaryAustin1** [reply, marked as answer]
That would only work for insert.  For select policy I assumed you were not limiting it.  You can't do a select on the table itself in hte select policy without moving all the of the select code to a security definer function.  It will keep calling itself over and over.

**Reply by GaryAustin1** [reply]
https://github.com/orgs/supabase/discussions/4509
If you search for recursion here in discussions you will find lots of hits.

**Reply by jonz-tech** [reply]
you are right! it work for insert, but when select policy is : auth.uid() = uid then infinite recursion detected!
I had search in supabase repo issue, but not search in discussions. I am sorry for not find that result! 
thank's for your help!

**Comment by jonz-tech** [comment]
> https://github.com/orgs/supabase/discussions/4509 If you search for recursion here in discussions you will find lots of hits.


sorry!
i try the definer function to advoid recursion, but it seems work in 2021 and not work now, recursion still occur!

here is my definer funtion: 
`
CREATE OR REPLACE FUNCTION fn_member_gids_by_uid(user_id uuid)
RETURNS SETOF BIGINT 
STABLE
SECURITY DEFINER
SET search_path = public
as $$
begin
  SELECT gid from tb_group_members gm where gm.uid = user_id;
end
$$ LANGUAGE plpgsql;

`

and for select policy:
`

create policy "query by auth uid" on "public"."tb_group_members" as permissive for select to authenticated using (
(
   gid IN (
         SELECT fn_member_gids_by_uid(
                  ( SELECT auth.uid() AS uid)) AS fn_member_gids_by_uid)
  )
);

`

**Reply by GaryAustin1** [reply]
I really have no idea what you are doing there.  But never pass in a UUID for the user to an RLS function.   Just use auth.uid() in the function.
